app/client/ui/Pages.ts

```diff
@@ -67,19 +68,27 @@ function buildDomFromTable(pagesTable: MetaTableModel<PageRec>, activeDoc: Grist
   }
 
   const {isReadonly} = activeDoc;
-  const pageName = pagesTable.rowModels[pageId].view.peek().name;
-  const viewId = pagesTable.rowModels[pageId].view.peek().id.peek();
+  const pageRec = pagesTable.rowModels[pageId];
+  const viewRec = pageRec.view.peek();
+  const pageName = viewRec.name;
+  const viewId = viewRec.id.peek();
+  const {docModel} = activeDoc;
 
-  const actions: PageActions = {
+  const options: PageOptions = {
     onRename: (newName: string) => newName.length && pageName.saveOnly(newName),
     onRemove: () => removeView(activeDoc, viewId, pageName.peek()),
     onDuplicate: () => buildDuplicatePageDialog(activeDoc, pageId),
     // Can't remove last visible page
-    isRemoveDisabled: () => activeDoc.docModel.visibleDocPages.peek().length <= 1,
-    isReadonly
+    isRemoveDisabled: () => docModel.visibleDocPages.peek().length <= 1,
+    isReadonly,
+    isCollapsed: pageRec.isCollapsed,
+    onCollapse: (value) => pageRec.isCollapsed.set(value),
+    isCollapsedByDefault: pageRec.isCollapsedByDefault,
+    onCollapseByDefault: (value) => pageRec.setAndSaveCollapsed(value),
+    hasSubPages: () => !docModel.leafPageIds.peek().has(pageRec.getRowId()),
```

Comment:
With my earlier suggestion, you can replace parameter `hidden: boolean` with `item: TreeITemRecord`, and that's enough to calculate `hasSubPages`:
```suggestion
    hasSubPages: () => item.children().get().length > 0,
```
(`hidden` is still available as `item.hidden`)